### PR TITLE
CI: On Travis matrix "OPTIMIZE=-OO" flag ignored

### LINKED ISSUE #7914: TravisCI not failing when it should for -OO run
See https://github.com/scipy/scipy/pull/7900#pullrequestreview-64708732. The PR had code that should be failing with ``-OO`` (use of ``__doc__`` without checking ``if __doc__ is not None``), but it didn't. Needs investigating. Could be a 2.7 vs. 3.x thing perhaps.

### PR DESCRIPTION
**Pull request by ilayn**
In order to address #9373, we need a working (in this case failing) build however Travis doesn't seem to enable the `-OO` flag. This PR is for testing different options in the yaml file.

Closes #7914

### CODE REVIEW

**Comment by ev-br** [comment]
I wonder if pytest respects the PYTHINOPTIMIZE env variable.

**Comment by ilayn** [comment]
Yes indeed that is my suspicion hence I've opened up the issue https://github.com/pytest-dev/pytest/issues/4189

**Comment by ilayn** [comment]
`pytest` indeed respects the env variable or `-OO` flag but the triggered load scheduler workers don't. I've asked about that in the linked issue.

**Comment by ilayn** [comment]
It seems that NumPy 1.8.2 back then had a `__doc__.split('\n')` clause in its `setup.py` so even we figure out our own build chain compatible with `-OO` we have to modify the incoming NumPy which does not seem to be worth it. 

Nevertheless I'll try to wrap up the pieces such that at least personal builds can go through with the newer versions of NumPy.

**Comment by rgommers** [comment]
It just needs to work for one CI config, which will use a recent NumPy. NumPy is responsible for its own -OO CI, no need for us to worry about older versions of that.

**Comment by ilayn** [comment]
Travis is now failing properly and predictably on the docstring absence. [These lines ](https://travis-ci.org/scipy/scipy/jobs/445880408#L1171)from build output shows that the three workers fired up with `-OO` flag

<details>

```Shell
271.44s$ python -u runtests.py -g -m $TESTMODE $COVERAGE $USE_WHEEL_BUILD -- -rfEX -n 3 2>&1 | tee runtests.log
WARNING: assertions not in test modules or plugins will be ignored because assert statements are not executed by the underlying Python interpreter (are you using python -O?)
============================= test session starts ==============================
platform linux -- Python 3.4.6, pytest-3.9.2, py-1.7.0, pluggy-0.8.0
rootdir: /home/travis/build/scipy/scipy, inifile: pytest.ini
plugins: xdist-1.23.2, forked-0.2, faulthandler-1.5.0
gw0 I / gw1 I / gw2 I
WARNING: assertions not in test modules or plugins will be ignored because assert statements are not executed by the underlying Python interpreter (are you using python -O?)
WARNING: assertions not in test modules or plugins will be ignored because assert statements are not executed by the underlying Python interpreter (are you using python -O?)
WARNING: assertions not in test modules or plugins will be ignored because assert statements are not executed by the underlying Python interpreter (are you using python -O?)
gw0 [12883] / gw1 [12883] / gw2 [12883]
scheduling tests via LoadScheduling
......................................s................................. [  0%]
.........................s.............................................. [  1%]
................s....................................................... [  1%]
........................................................................ [  2%]
........................................................F............... [  2%]
............s..............................................x............ [  3%]
........................................................................ [  3%]
........................................................................ [  4%]
........................................................................ [  5%]

...
```

</details>

I think together with #9373 this would solve the issue. I'll wait a bit and merge them if there is no objection. 

I've touched very little to all involved files and for the `runtests.py` they are all PEP8 cleanup I did while I was experimenting with it.

**Comment by larsoner** [comment]
> I'll wait a bit and merge them if there is no objection.

I'd rather just have you cherry pick the commit from the PR that will make everything green again. Then we don't have to merge this red PR, which is always a risk even if you have the fix immediately lined up. Better to prove in this PR that the fix works, and find out here whether or not it does, than do this experimentation in master.

**Comment by ilayn** [comment]
I did the local experiments but indeed that would be less error prone.

**Comment by andyfaff** [comment]
One thing I've found useful on iterating CI changes is to enable TravisCI on your own scipy fork. That way you can make as many changes as you like, and the CI machinery will do the requisite testing. Then, once everything works you can submit a PR. It'll require you signing up for a Travis account.

**Comment by ilayn** [comment]
After cherrypicking the only failures on Travis are the usual suspects that have been creeping up for some time now. Next stop is actually figuring out those fails.

**Comment by larsoner** [comment]
So we should merge this, then fix those in a separate PR, right?

**Comment by ilayn** [comment]
Yes, this one is done in my opinion.

**Review by larsoner — COMMENTED** [review]
Other than one comment-related comment, +1 for merge

File: .travis.yml [hunk 1]
```diff
@@ -46,13 +46,13 @@ matrix:
...
         - NUMPYSPEC="numpy==1.8.2"
         - USE_SDIST=1
-        - OPTIMIZE=-OO
     - python: 3.4
       env:
         - TESTMODE=fast
         - COVERAGE=
         - USE_WHEEL=1
         - NUMPYSPEC="--upgrade numpy"
+        - PYTHONOPTIMIZE=2
```
[the remark this answers is not in the document]
Comment (ilayn) [reply .travis.yml]: Added a brief comment.

**Review by larsoner — APPROVED** [review]

**Pull request merged**